File: runtimes/python/helix_rt/balancer.py
class RoundRobinBalancer:
    def __init__(self, clients=None):
        self.clients = clients or []
        self.index = 0

    def register(self, client):
        self.clients.append(client)

    def next(self):
        if not self.clients:
            raise Exception("no clients registered in balancer")
        client = self.clients[self.index]
        self.index = (self.index + 1) % len(self.clients)
        return client

File: runtimes/node/src/balancer.ts
export interface Client {
    invoke(path: string, request: any): Promise<any>;
}

export class RoundRobinBalancer {
    private clients: Client[] = [];
    private index: number = 0;

    constructor(clients: Client[] = []) {
        this.clients = clients;
    }

    public register(client: Client) {
        this.clients.push(client);
    }

    public next(): Client {
        if (this.clients.length === 0) {
            throw new Error("no clients registered in balancer");
        }
        const client = this.clients[this.index];
        this.index = (this.index + 1) % this.clients.length;
        return client;
    }
}
